go-controller/pkg/ovn/controller/apbroute/external_controller.go

```diff
@@ -128,45 +153,92 @@ func newExternalPolicyManager(
 	netClient networkClient) *externalPolicyManager {
 
 	m := externalPolicyManager{
-		stopCh:                 stopCh,
-		routeLister:            routeLister,
-		podLister:              podLister,
-		namespaceLister:        namespaceLister,
-		namespaceInfoSyncCache: syncmap.NewSyncMap[*namespaceInfo](),
-		routePolicySyncCache:   syncmap.NewSyncMap[*adminpolicybasedrouteapi.AdminPolicyBasedExternalRoute](),
-		netClient:              netClient,
+		stopCh:                         stopCh,
+		routeLister:                    routeLister,
+		podLister:                      podLister,
+		namespaceLister:                namespaceLister,
+		namespaceInfoSyncCache:         syncmap.NewSyncMap[*namespaceInfo](),
+		routePolicySyncCache:           syncmap.NewSyncMap[*routeInfo](),
+		netClient:                      netClient,
+		mutexRoutePolicyCachePopulated: &sync.Mutex{},
 	}
 
 	return &m
 }
 
+func (m *externalPolicyManager) setRoutePolicyCacheAsPopulated() {
+	m.mutexRoutePolicyCachePopulated.Lock()
+	defer m.mutexRoutePolicyCachePopulated.Unlock()
+	m.routePolicyCachePopulated = true
+}
+
+func (m *externalPolicyManager) isRoutePolicyCachePopulated() bool {
+	m.mutexRoutePolicyCachePopulated.Lock()
+	defer m.mutexRoutePolicyCachePopulated.Unlock()
+	return m.routePolicyCachePopulated
+}
+
 // getRoutePolicyFromCache retrieves the cached value of the policy if it exists in the cache, as well as locking the key in case it exists.
-func (m *externalPolicyManager) getRoutePolicyFromCache(policyName string) (*adminpolicybasedrouteapi.AdminPolicyBasedExternalRoute, bool) {
+func (m *externalPolicyManager) getRoutePolicyFromCache(policyName string) (*adminpolicybasedrouteapi.AdminPolicyBasedExternalRoute, bool, bool) {
 	var (
-		policy *adminpolicybasedrouteapi.AdminPolicyBasedExternalRoute
-		found  bool
+		policy                   *adminpolicybasedrouteapi.AdminPolicyBasedExternalRoute
+		found, markedForDeletion bool
 	)
 	_ = m.routePolicySyncCache.DoWithLock(policyName, func(policyName string) error {
-		policy, found = m.routePolicySyncCache.Load(policyName)
+		ri, f := m.routePolicySyncCache.Load(policyName)
+		if !f {
+			return nil
+		}
+		found = f
+		policy = ri.policy
```

Comment:
so I think this is unsafe. The value in the cache (along with namespaceinfo cache is a pointer). So when you set this, unlock the map, and return. You are returning the poitner to the thing in the map, which something else is going to read/modify outside of lock. I think you need to make a copy of the value of the pointer and return that if you are going to use it outside of the lock. Might want to check namespaceinfo cache to see if that happens there too.